runtime/src/main/cpp/Memory.cpp

```diff
@@ -2025,25 +2029,32 @@ OBJ_GETTER(initSharedInstance,
   }
 #endif  // KONAN_NO_EXCEPTIONS
 #else  // KONAN_NO_THREADS
-  ObjHeader* value = *localLocation;
-  if (value != nullptr) RETURN_OBJ(value);
+  auto it = memoryState->initializingSingletons.find(location);
+  if (it != memoryState->initializingSingletons.end()) {
+    RETURN_OBJ(it->second);
+  }
 
   ObjHeader* initializing = reinterpret_cast<ObjHeader*>(1);
 
   // Spin lock.
+  ObjHeader* value = nullptr;
   while ((value = __sync_val_compare_and_swap(location, nullptr, initializing)) == initializing);
   if (value != nullptr) {
     // OK'ish, inited by someone else.
     RETURN_OBJ(value);
   }
   ObjHeader* object = AllocInstance(typeInfo, OBJ_RESULT);
-  UpdateHeapRef(localLocation, object);
+  auto insertIt = memoryState->initializingSingletons.insert({location, object});
+  RuntimeCheck(insertIt.second, "object cannot be assigned twice into initializingSingletons");
+  addHeapRef(object);
```

Comment:
Is this actually required?

Reply:
I'm not sure, but what's protecting this reference (and all that it references) from being garbage collected while (potentially long) initialization is taking place?

Reply:
The allocation [above](https://github.com/JetBrains/kotlin-native/pull/4019/files#diff-c7d26c58627c81c35c99320f067b2752R2046) actually creates a stack reference. It is alive at least until the return from the function.